README.md
StatsD
======

A network daemon for aggregating statistics (counters and timers), rolling them up, then sending them to graphing services such as [graphite][graphite] or [Librato Metrics][metrics].

We ([Etsy][etsy]) [blogged][blog post] about how it works and why we created it.


Concepts
--------

* *buckets*
  Each stat is in its own "bucket". They are not predefined anywhere. Buckets can be named anything that will translate to Graphite (periods make folders, etc)

* *values*
  Each stat will have a value. How it is interpreted depends on modifiers

* *flush*
  After the flush interval timeout (default 10 seconds), stats are munged and sent over to Graphite.

Counting
--------

    gorets:1|c

This is a simple counter. Add 1 to the "gorets" bucket. It stays in memory until the flush interval `config.flushInterval`.

Timing
------

    glork:320|ms

The glork took 320ms to complete this time. StatsD figures out 90th percentile, average (mean), lower and upper bounds for the flush interval.  The percentile threshold can be tweaked with `config.percentThreshold`.

Sampling
--------

    gorets:1|c|@0.1

Tells StatsD that this counter is being sent sampled every 1/10th of the time.

Debugging
---------

There are additional config variables available for debugging:

* `debug` - log exceptions and periodically print out information on counters and timers
* `debugInterval` - interval for printing out information on counters and timers
* `dumpMessages` - print debug info on incoming messages

For more information, check the `exampleConfig.js`.

Logging
-------

Nominally the output on stdout is kept clean and all system events are logged through syslog.
If `debug` is set, all output is redirected to stdout/stderr accordingly.

Guts
----

* [UDP][udp]
  Client libraries use UDP to send information to the StatsD daemon.

* [NodeJS][node]

Statsd can submit data to multiple graphing services. Currently these include:

* [Graphite][graphite]

* [Librato Metrics][metrics]

Graphing Services
-----------------

**Graphite**
Graphite is an open source highly scalable real-time graphing system. Graphite uses "schemas" to define the different round robin datasets it houses (analogous to RRAs in rrdtool). Here's what Etsy is using for the stats databases:

    [stats]
    priority = 110
    pattern = ^stats\..*
    retentions = 10:2160,60:10080,600:262974

That translates to:

* 6 hours of 10 second data (what we consider "near-realtime")
* 1 week of 1 minute data
* 5 years of 10 minute data

This has been a good tradeoff so far between size-of-file (round robin databases are fixed size) and data we care about. Each "stats" database is about 3.2 megs with these retentions.

**Librato Metrics**
Librato Metrics is a hosted graphing service that eliminates the need to set up and run your own graphing server. Submitting data to [Librato Metrics][metrics] requires a username and API token. You can find your API token in your [account settings](https://metrics.librato.com/account).

TCP Stats Interface
-------------------

A really simple TCP management interface is available by default on port 8126 or overriden in the configuration file. Inspired by the memcache stats approach this can be used to monitor a live statsd server.  You can interact with the management server by telnetting to port 8126, the following commands are available:

* stats - some stats about the running server
* counters - a dump of all the current counters
* timers - a dump of the current timers

The stats output currently will give you:

* uptime: the number of seconds elapsed since statsd started
* graphite.last_flush: the number of seconds elapsed since the last successful flush to graphite
* graphite.last_exception: the number of seconds elapsed since the last exception thrown whilst flushing to graphite
* messages.last_msg_seen: the number of elapsed seconds since statsd received a message
* messages.bad_lines_seen: the number of bad lines seen since startup

Installation and Configuration
------------------------------

 * Install node.js
 * Install the 'base64', 'async' and 'underscore' packages via [npm][npm]
 * Clone the project
 * Create a config file from exampleConfig.js and put it somewhere
 * Add connection info for your graphing service to exampleConfig.js
 * Start the Daemon:

```
node stats.js /path/to/config
```

Tests
-----

A test framework has been added using node-unit and some custom code to start and manipulate statsd. Please add tests under test/ for any new features or bug fixes encountered. Testing a live server can be tricky, attempts were made to eliminate race conditions but it may be possible to encounter a stuck state. If doing dev work, a `killall node` will kill any stray test servers in the background (don't do this on a production machine!).

Tests can be executd with `./run_tests.sh`.

Inspiration
-----------

StatsD was inspired (heavily) by the project (of the same name) at Flickr. Here's a post where Cal Henderson described it in depth:
[Counting and timing](http://code.flickr.com/blog/2008/10/27/counting-timing/). Cal re-released the code recently: [Perl StatsD](https://github.com/iamcal/Flickr-StatsD)


Contribute
---------------------

You're interested in contributing to StatsD? *AWESOME*. Here are the basic steps:

fork StatsD from here: http://github.com/etsy/statsd

1. Clone your fork
2. Hack away
3. If you are adding new functionality, document it in the README
4. If necessary, rebase your commits into logical chunks, without errors
5. Push the branch up to GitHub
6. Send a pull request to the etsy/statsd project.

We'll do our best to get your changes in!

[graphite]: http://graphite.wikidot.com
[metrics]: https://metrics.librato.com
[etsy]: http://www.etsy.com
[blog post]: http://codeascraft.etsy.com/2011/02/15/measure-anything-measure-everything/
[node]: http://nodejs.org
[udp]: http://enwp.org/udp
[npm]: http://npmjs.org


Contributors
-----------------

In lieu of a list of contributors, check out the commit history for the project:
http://github.com/etsy/statsd/commits/master


debian/changelog
statsd (0.0.2) unstable; urgency=low

  * Bumped version number

 -- Trung Duc Tran <trung@estately.com> Tue, 2 Oct 2012 22:07:34 +0200

statsd (0.0.1) unstable; urgency=low

  * Initial rdio packaging.

 -- Manish Singh <manish@rd.io>  Mon, 21 Mar 2011 23:04:52 -0700


test/librato_tests.js
var fs           = require('fs'),
    temp         = require('temp'),
    spawn        = require('child_process').spawn,
    urlparse     = require('url').parse,
    _            = require('underscore'),
    dgram        = require('dgram'),
    qsparse      = require('querystring').parse,
    http         = require('http');


var writeconfig = function(text,worker,cb,obj){
  temp.open({suffix: '-statsdconf.js'}, function(err, info) {
    if (err) throw err;
    fs.write(info.fd, text);
    fs.close(info.fd, function(err) {
      if (err) throw err;
      worker(info.path,cb,obj);
    });
  });
}

var array_contents_are_equal = function(first,second){
  var intlen = _.intersection(first,second).length;
  var unlen = _.union(first,second).length;
  return (intlen == unlen) && (intlen == first.length);
}

var statsd_send = function(data,sock,host,port,cb){
  send_data = new Buffer(data);
  sock.send(send_data,0,send_data.length,port,host,function(err,bytes){
    if (err) {
      throw err;
    }
    cb();
  });
}

// keep collecting data until a specified timeout period has elapsed
// this will let us capture all data chunks so we don't miss one
var collect_for = function(server,timeout,cb){
  var received = [];
  var in_flight = 0;
  var start_time = new Date().getTime();
  var collector = function(req,res){
    res.writeHead(204);
    res.end();
    in_flight += 1;
    var body = '';
    req.on('data',function(data){ body += data; });
    req.on('end',function(){
      received = received.concat(body);
      in_flight -= 1;
      if((in_flight < 1) && (new Date().getTime() > (start_time + timeout))){
          server.removeListener('request',collector);
          cb(received);
      }
    });
  }

  setTimeout(function (){
    server.removeListener('request',collector);
    if((in_flight < 1)){
      cb(received);
    }
  },timeout);

  server.on('request',collector);
}

module.exports = {
  setUp: function (callback) {
    this.testport = 31337;

    this.myflush = 200;
    var configfile = "{graphService: \"librato-metrics\"\n\
               ,  batch: 200 \n\
               ,  flushInterval: " + this.myflush + " \n\
               ,  port: 8125\n\
               ,  dumpMessages: false \n\
               ,  debug: true\n\
               ,  libratoUser: \"test@librato.com\"\n\
               ,  libratoSnap: 10\n\
               ,  libratoApiKey: \"fakekey\"\n\
               ,  libratoHost: \"http://127.0.0.1:" + this.testport + "\"}";

    this.acceptor = http.createServer();
    this.acceptor.listen(this.testport);
    this.sock = dgram.createSocket('udp4');

    this.server_up = true;
    this.ok_to_die = false;
    this.exit_callback_callback = process.exit;

    writeconfig(configfile,function(path,cb,obj){
      obj.path = path;
      obj.server = spawn('node',['stats.js', path]);
      obj.exit_callback = function (code) {
        obj.server_up = false;
        if(!obj.ok_to_die){
          console.log('node server unexpectedly quit with code: ' + code);
          process.exit(1);
        }
        else {
          obj.exit_callback_callback();
        }
      };
      obj.server.on('exit', obj.exit_callback);
      obj.server.stderr.on('data', function (data) {
        console.log('stderr: ' + data.toString().replace(/\n$/,''));
      });
      /*
      obj.server.stdout.on('data', function (data) {
        console.log('stdout: ' + data.toString().replace(/\n$/,''));
      });
      */
      obj.server.stdout.on('data', function (data) {
        // wait until server is up before we finish setUp
        if (data.toString().match(/server is up/)) {
          cb();
        }
      });

    },callback,this);
  },
  tearDown: function (callback) {
    this.sock.close();
    this.acceptor.close();
    this.ok_to_die = true;
    if(this.server_up){
      this.exit_callback_callback = callback;
      this.server.kill();
    } else {
      callback();
    }
  },

  send_well_formed_posts: function (test) {
    test.expect(5);

    // we should integrate a timeout into this
    this.acceptor.once('request',function(req,res){
        res.writeHead(204);
        res.end();
        test.equals(req.method,'POST');
        var uri_parts = urlparse(req.url);
        test.equals(uri_parts["pathname"],'/v1/metrics.json')
        var body = '';
        req.on('data',function(data){ body += data; });
        req.on('end',function(){
          try {
              var post = JSON.parse(body);
          } catch (e) {
            test.ok(false,"string sent was not valid JSON: " + e);
            test.done();
            return;
          }
          test.ok(true);
          var message_keys = ['measure_time','gauges'];
          test.ok(array_contents_are_equal(_.keys(post),message_keys),"JSON must only have: [" + message_keys + "], received: [" + _.keys(post) + "]");

          if(_.include(_.keys(post),'gauges') && _.include(_.keys(post['gauges']),'numStats') &&
             _.include(_.keys(post['gauges']['numStats']),'value')){
             test.equals(post['gauges']['numStats']['value'],0);
          } else {
            test.ok('false', 'API does not send numStats properly');
          }
          test.done();
        });
    });
  },

  counts_are_valid: function (test) {
    test.expect(3);

    var testvalue = 100;
    var me = this;
    this.acceptor.once('request',function(req,res){
      res.writeHead(204);
      res.end();
      statsd_send('a_test_value:' + testvalue + '|ms',me.sock,'127.0.0.1',8125,function(){
          collect_for(me.acceptor,me.myflush*2,function(strings){
            test.ok(strings.length > 0,'should receive some data');
            var hashes = _.map(strings,function(str){ return JSON.parse(str); });
            var numstat_test = function(post){
              return _.include(_.keys(post),'gauges') && _.include(_.keys(post['gauges']),'numStats') &&
              _.include(_.keys(post['gauges']['numStats']),'value') && 
              (post['gauges']['numStats']['value'] == 1);
            };
            test.ok(_.any(hashes,numstat_test), 'numStats should be 1');

            var testvalue_test = function(post){
              return _.include(_.keys(post),'gauges') && _.include(_.keys(post['gauges']),'a_test_value') &&
              _.include(_.keys(post['gauges']['a_test_value']),'sum') && 
              (post['gauges']['a_test_value']['sum'] == testvalue);
            };
            test.ok(_.any(hashes,testvalue_test), 'testvalue should be ' + testvalue);
            test.done();
          });
      });
    });
  }
}


test/graphite_tests.js
var fs           = require('fs'),
    net          = require('net'),
    temp         = require('temp'),
    spawn        = require('child_process').spawn,
    sys          = require('util'),
    urlparse     = require('url').parse,
    _            = require('underscore'),
    dgram        = require('dgram'),
    qsparse      = require('querystring').parse,
    http         = require('http');


var writeconfig = function(text,worker,cb,obj){
  temp.open({suffix: '-statsdconf.js'}, function(err, info) {
    if (err) throw err;
    fs.write(info.fd, text);
    fs.close(info.fd, function(err) {
      if (err) throw err;
      worker(info.path,cb,obj);
    });
  });
}

var array_contents_are_equal = function(first,second){
  var intlen = _.intersection(first,second).length;
  var unlen = _.union(first,second).length;
  return (intlen == unlen) && (intlen == first.length);
}

var statsd_send = function(data,sock,host,port,cb){
  send_data = new Buffer(data);
  sock.send(send_data,0,send_data.length,port,host,function(err,bytes){
    if (err) {
      throw err;
    }
    cb();
  });
}

// keep collecting data until a specified timeout period has elapsed
// this will let us capture all data chunks so we don't miss one
var collect_for = function(server,timeout,cb){
  var received = [];
  var in_flight = 0;
  var start_time = new Date().getTime();
  var collector = function(req,res){
    in_flight += 1;
    var body = '';
    req.on('data',function(data){ body += data; });
    req.on('end',function(){
      received = received.concat(body.split("\n"));
      in_flight -= 1;
      if((in_flight < 1) && (new Date().getTime() > (start_time + timeout))){
          server.removeListener('request',collector);
          cb(received);
      }
    });
  }

  setTimeout(function (){
    server.removeListener('connection',collector);
    if((in_flight < 1)){
      cb(received);
    }
  },timeout);

  server.on('connection',collector);
}

module.exports = {
  setUp: function (callback) {
    this.testport = 31337;
    this.myflush = 200;
    var configfile = "{graphService: \"graphite\"\n\
               ,  batch: 200 \n\
               ,  flushInterval: " + this.myflush + " \n\
               ,  port: 8125\n\
               ,  dumpMessages: false \n\
               ,  debug: false\n\
               ,  graphitePort: " + this.testport + "\n\
               ,  graphiteHost: \"127.0.0.1\"}";

    this.acceptor = net.createServer();
    this.acceptor.listen(this.testport);
    this.sock = dgram.createSocket('udp4');

    this.server_up = true;
    this.ok_to_die = false;
    this.exit_callback_callback = process.exit;

    writeconfig(configfile,function(path,cb,obj){
      obj.path = path;
      obj.server = spawn('node',['stats.js', path]);
      obj.exit_callback = function (code) {
        obj.server_up = false;
        if(!obj.ok_to_die){
          console.log('node server unexpectedly quit with code: ' + code);
          process.exit(1);
        }
        else {
          obj.exit_callback_callback();
        }
      };
      obj.server.on('exit', obj.exit_callback);
      obj.server.stderr.on('data', function (data) {
        console.log('stderr: ' + data.toString().replace(/\n$/,''));
      });
      /*
      obj.server.stdout.on('data', function (data) {
        console.log('stdout: ' + data.toString().replace(/\n$/,''));
      });
      */
      obj.server.stdout.on('data', function (data) {
        // wait until server is up before we finish setUp
        if (data.toString().match(/server is up/)) {
          cb();
        }
      });

    },callback,this);
  },
  tearDown: function (callback) {
    this.sock.close();
    this.acceptor.close();
    this.ok_to_die = true;
    if(this.server_up){
      this.exit_callback_callback = callback;
      this.server.kill();
    } else {
      callback();
    }
  },

  send_well_formed_posts: function (test) {
    test.expect(2);

    // we should integrate a timeout into this
    this.acceptor.once('connection',function(c){
      var body = '';
      c.on('data',function(d){ body += d; });
      c.on('end',function(){
        var rows = body.split("\n");
        var entries = _.map(rows, function(x) {
          var chunks = x.split(' ');
          var data = {};
          data[chunks[0]] = chunks[1];
          return data;
        });
        test.ok(_.include(_.map(entries,function(x) { return _.keys(x)[0] }),'statsd.numStats'),'graphite output includes numStats');
        test.equal(_.find(entries, function(x) { return _.keys(x)[0] == 'statsd.numStats' })['statsd.numStats'],0);
        test.done();
      });
    });
  },

  timers_are_valid: function (test) {
    test.expect(3);

    var testvalue = 100;
    var me = this;
    this.acceptor.once('connection',function(c){
      statsd_send('a_test_value:' + testvalue + '|ms',me.sock,'127.0.0.1',8125,function(){
          collect_for(me.acceptor,me.myflush*2,function(strings){
            test.ok(strings.length > 0,'should receive some data');
            var hashes = _.map(strings, function(x) {
              var chunks = x.split(' ');
              var data = {};
              data[chunks[0]] = chunks[1];
              return data;
            });
            var numstat_test = function(post){
              var mykey = 'statsd.numStats';
              return _.include(_.keys(post),mykey) && (post[mykey] == 1);
            };
            test.ok(_.any(hashes,numstat_test), 'statsd.numStats should be 1');

            var testtimervalue_test = function(post){
              var mykey = 'stats.timers.a_test_value.mean';
              return _.include(_.keys(post),mykey) && (post[mykey] == testvalue);
            };
            test.ok(_.any(hashes,testtimervalue_test), 'stats.timers.a_test_value.mean should be ' + testvalue);

            test.done();
          });
      });
    });
  },

  counts_are_valid: function (test) {
    test.expect(4);

    var testvalue = 100;
    var me = this;
    this.acceptor.once('connection',function(c){
      statsd_send('a_test_value:' + testvalue + '|c',me.sock,'127.0.0.1',8125,function(){
          collect_for(me.acceptor,me.myflush*2,function(strings){
            test.ok(strings.length > 0,'should receive some data');
            var hashes = _.map(strings, function(x) {
              var chunks = x.split(' ');
              var data = {};
              data[chunks[0]] = chunks[1];
              return data;
            });
            var numstat_test = function(post){
              var mykey = 'statsd.numStats';
              return _.include(_.keys(post),mykey) && (post[mykey] == 1);
            };
            test.ok(_.any(hashes,numstat_test), 'statsd.numStats should be 1');

            var testavgvalue_test = function(post){
              var mykey = 'stats.a_test_value';
              return _.include(_.keys(post),mykey) && (post[mykey] == (testvalue/(me.myflush / 1000)));
            };
            test.ok(_.any(hashes,testavgvalue_test), 'stats.a_test_value should be ' + (testvalue/(me.myflush / 1000)));

            var testcountvalue_test = function(post){
              var mykey = 'stats_counts.a_test_value';
              return _.include(_.keys(post),mykey) && (post[mykey] == testvalue);
            };
            test.ok(_.any(hashes,testcountvalue_test), 'stats_counts.a_test_value should be ' + testvalue);

            test.done();
          });
      });
    });
  }
}


stats.js
versionstring = "statsd-librato/1.3"

var dgram      = require('dgram')
  , sys        = require('util')
  , net        = require('net')
  , config     = require('./config')
  , _          = require('underscore')
  , async      = require('async')
  , url_parse  = require('url').parse
  , https      = require('https')
  , http       = require('http')
  , syslog     = require('node-syslog');

var logger = function(severity,message){
  if(severity < syslog.LOG_NOTICE){
    console.error(message);
  } else {
    console.log(message);
  }
}

var counters = {};
var timers = {};
var gauges = {};
var debugInt, flushInt, server, mgmtServer;
var startup_time = Math.round(new Date().getTime() / 1000);

var globalstats = {
  graphite: {
    last_flush: startup_time,
    last_exception: startup_time
  },
  messages: {
    last_msg_seen: startup_time,
    bad_lines_seen: 0,
  }
};

process.on('uncaughtException', function (err) {
  logger(syslog.LOG_ERR, 'Caught exception: ' + err.stack);
  logger(syslog.LOG_ERR, 'Aborting server...');
  process.exit(1);
});

config.configFile(process.argv[2], function (config, oldConfig) {
  if (config.debug){
    sys.log('logging to stdout/err');
  } else {
    syslog.init("node-syslog", syslog.LOG_PID | syslog.LOG_ODELAY, syslog.LOG_LOCAL0);
    sys.log('logging to syslog');
    logger = syslog.log;
  }

  function graphServiceIs(name){
    return (config.graphService) && (config.graphService == name);
  }

  if (!config.debug && debugInt) {
    clearInterval(debugInt);
    debugInt = false;
  }

  if (config.debug) {
    if (debugInt !== undefined) { clearInterval(debugInt); }
    debugInt = setInterval(function () {
      logger(syslog.LOG_INFO, "Counters:\n" + sys.inspect(counters) + "\nGauges:\n" + sys.inspect(gauges) + "\nTimers:\n" + sys.inspect(timers));
    }, config.debugInterval || 10000);
  }

  if (server === undefined) {
    server = dgram.createSocket('udp4', function (msg, rinfo) {
      var msgStr = msg.toString().replace(/^\s+|\s+$/g,"").replace(/\u0000/g, '');
      if (msgStr.length == 0) {
        if (config.debug) {
          logger(syslog.LOG_DEBUG, 'No messsages.');
        }
        return;
      }
      if (config.dumpMessages) {
        logger(syslog.LOG_INFO, 'Messages: ' + msgStr);
      }
      var bits = msgStr.split(':');
      var key = '';
      key = bits.shift();

      if (bits.length == 0) {
        return;
      }

      for (var i = 0; i < bits.length; i++) {
        var sampleRate = 1;
        var fields = bits[i].split("|");
        if (fields[1] === undefined) {
            logger(syslog.LOG_ERR, 'Bad line: ' + fields);
            globalstats['messages']['bad_lines_seen']++;
            continue;
        }
        if (fields[1].trim() == "ms") {
          if (! timers[key]) {
            timers[key] = [];
          }
          timers[key].push(Number(fields[0] || 0));
        } else if (fields[1].trim() === "g") {
          gauges[key] = Number(fields[0] || 0);
        } else {
          if (fields[2] && fields[2].match(/^@([\d\.]+)/)) {
            sampleRate = Number(fields[2].match(/^@([\d\.]+)/)[1]);
          }
          if (! counters[key]) {
            counters[key] = 0;
          }
          counters[key] += Number(fields[0] || 1) * (1 / sampleRate);
        }
      }

      globalstats['messages']['last_msg_seen'] = Math.round(new Date().getTime() / 1000);
    });

    server.on("listening", function () {
      var address = server.address();
      logger(syslog.LOG_INFO, "statsd is running on " + address.address + ":" + address.port);
      sys.log("server is up");
    });

    mgmtServer = net.createServer(function(stream) {
      stream.setEncoding('ascii');

      stream.on('data', function(data) {
        var cmd = data.trim();

        switch(cmd) {
          case "help":
            stream.write("Commands: stats, counters, gauges, timers, quit\n\n");
            break;

          case "stats":
            var now    = Math.round(new Date().getTime() / 1000);
            var uptime = now - startup_time;

            stream.write("uptime: " + uptime + "\n");

            for (group in stats) {
              for (metric in stats[group]) {
                var val;

                if (metric.match("^last_")) {
                  val = now - stats[group][metric];
                }
                else {
                  val = stats[group][metric];
                }

                stream.write(group + "." + metric + ": " + val + "\n");
              }
            }
            stream.write("END\n\n");
            break;

          case "counters":
            stream.write(sys.inspect(counters) + "\n");
            stream.write("END\n\n");
            break;

          case "timers":
            stream.write(sys.inspect(timers) + "\n");
            stream.write("END\n\n");
            break;

          case "gauges":
            stream.write(sys.inspect(gauges) + "\n");
            stream.write("END\n\n");
            break;

          case "quit":
            stream.end();
            break;

          default:
            stream.write("ERROR\n");
            break;
        }

      });
    });

    server.bind(config.port || 8125);
    mgmtServer.listen(config.mgmt_port || 8126);

    var flushInterval = Number(config.flushInterval || 10000);

    flushInt = setInterval(function () {
      var stats = {};
      stats["gauges"] = {};
      stats["counters"] = {};
      var ts = Math.round(new Date().getTime() / 1000);
      var numStats = 0;
      var key;

      for (key in counters) {
        var stat;
        stat = stats["counters"];

        var value = counters[key];
        stat[key] = {};
        stat[key]["value"] = value;

        counters[key] = 0;

        numStats += 1;
      }

      for (key in gauges) {
        var stat;
        stat = stats["gauges"];

        var value = gauges[key];
        stat[key] = {};
        stat[key]["value"] = value;

        delete gauges[key];

        numStats += 1;
      }

      for (key in timers) {
        if (timers[key].length > 0) {
          var pctThreshold = config.percentThreshold || 90;
          var values = timers[key].sort(function (a,b) { return a-b; });
          var count = values.length;
          var min = values[0];
          var max = values[count - 1];

          var mean = min;
          var maxAtThreshold = max;

          if (count > 1) {
            var thresholdIndex = Math.round(((100 - pctThreshold) / 100) * count);
            var numInThreshold = count - thresholdIndex;
            values_sliced = values.slice(0, numInThreshold);
            maxAtThreshold = values_sliced[numInThreshold - 1];
            // average the remaining timings
            var sum = 0;
            for (var i = 0; i < numInThreshold; i++) {
              sum += values_sliced[i];
            }
            mean = sum / numInThreshold;
          }

          var sum = 0;
          var sumOfSquares = 0;
          for (var i = 0; i < count; i++) {
            sum += values[i];
            sumOfSquares += values[i] * values[i];
          }

          timers[key] = [];
          stats["gauges"][key] = {};
          stats["gauges"][key]["count"] = count;
          stats["gauges"][key]["sum"] = sum;
          stats["gauges"][key]["sum_squares"] = sumOfSquares;
          stats["gauges"][key]["min"] = min;
          stats["gauges"][key]["max"] = max;
          if (!graphServiceIs("librato-metrics")){
            stats["gauges"][key]["upper_" + pctThreshold] = maxAtThreshold;
            stats["gauges"][key]["mean"] = mean;
          }

          numStats += 1;
        }
      }

      stats["counters"]["numStats"] = {};
      stats["counters"]["numStats"]["value"] = numStats;

      var slicey = function(obj,slicelen){
        var slicecounter = 0;
        var groups = _.groupBy(obj,function (num){ var ret = Math.floor(slicecounter/slicelen); slicecounter += + 1; return ret;});
        return _.map(groups,function(k,v){ return k; });
      }

      function build_hash(type){
        return function(group){
          if (graphServiceIs("librato-metrics")){
            stat = stats["gauges"];
          }
          var hash = {};
          hash[type] = {};
          _.each(group,function(metric){
            hash[type][metric] = stats[type][metric];
          });
          if (graphServiceIs("librato-metrics")){
            hash["measure_time"] = ts
          }
          return hash;
        };
      }

      function hash_postprocess(inhash){
        if (graphServiceIs("librato-metrics")){
          var hash = {};
          hash["gauges"] = inhash["gauges"] || {};
          if (_.include(_.keys(inhash),"counters")) {
            _.each(_.keys(inhash["counters"]),function(metric){
                hash["gauges"][metric] = inhash["counters"][metric];
                });
          }
          snap = config.libratoSnap || 10;
          hash["measure_time"] = (ts - (ts%snap));
          if (config.libratoSource) { hash["source"] = config.libratoSource; }
          return hash;
        } else {
          return inhash;
        }
      }

      function build_string(hash,type){
          var stats_str ='';

          if (graphServiceIs("librato-metrics")){
            stats_str = JSON.stringify(hash);
          } else {
            for (key in hash[type]){
              if (graphServiceIs("librato-metrics")){
                k =       key.replace(/[^-.:_\w]+/, '_').substr(0,255)
              } else {
                k =       key
                          .replace(/\s+/g, '_')
                          .replace(/\//g, '-')
                          .replace(/[^a-zA-Z_\-0-9\.]/g, '');
              }

              var stat = hash[type][k];
              if (type == "counters"){
                if(k == 'numStats'){
                  stats_str += 'statsd.numStats ' + stat['value'] + ' ' + ts + "\n";
                } else if (stat["value"] > 0){
                  stats_str += ('stats.counters.' + k + ' ' + stat["value"] + ' ' + ts + "\n");
                }
              } else if (type == "gauges") {
                stats_str += ('stats.gauges.' + k + ' ' + stat['value'] + ' ' + ts + "\n");
              } else {
                for (s in stat){
                  stats_str += ('stats.timers.' + k + '.' + s + ' ' + stat[s] + ' ' + ts + "\n");
                }
              }
            }
          }
          return stats_str;
      };

      var slice_length = config.batch || 200;
      var ggroups = slicey(_.keys(stats["gauges"]),slice_length);
      var cgroups = slicey(_.keys(stats["counters"]),slice_length);
      var ghashes = _.map(ggroups,build_hash("gauges"));
      var chashes = _.map(cgroups,build_hash("counters"));
      var combined_hashes = ghashes.concat(chashes);

      var logerror = function(e){
        if (e){
          globalstats['graphite']['last_exception'] = Math.round(new Date().getTime() / 1000);
          if(config.debug) {
            logger(syslog.LOG_DEBUG, e);
          }
        }
      }

      var submit_to_librato = function(stats_str,retry){
        var parsed_host = url_parse(config.libratoHost || 'https://metrics-api.librato.com');
        var options = {
          host: parsed_host["hostname"],
          port: parsed_host["port"] || 443,
          path: '/v1/metrics.json',
          method: 'POST',
          headers: {
            "Authorization": 'Basic ' + new Buffer(config.libratoUser + ':' + config.libratoApiKey).toString('base64'),
            "Content-Length": stats_str.length,
            "Content-Type": "application/json",
            "User-Agent" : versionstring
          }
        };

        var proto = http;
        if ((parsed_host["protocol"] || 'http:').match(/https/)){
          proto = https;
        }
        var req = proto.request(options, function(res) {
          if(res.statusCode != 204){
            res.on('data', function(d){
              var errdata = "HTTP " + res.statusCode + ": " + d;
              if (retry){
                if (config.debug) { console.log("received error " + res.statusCode + " connecting to Librato, retrying... "); }
                setTimeout(function(){
                  submit_to_librato(stats_str,false);
                }, Math.floor(flushInterval/2) + 100);
              } else {
                logger(syslog.LOG_CRIT, "Error connecting to Librato!\n" + errdata);
              }
            });
          }
        });
        req.write(stats_str);
        req.end();
        globalstats['graphite']['last_flush'] = Math.round(new Date().getTime() / 1000);
        req.on('error', function(errdata) {
            if (retry){
              setTimeout(function(){
                submit_to_librato(stats_str,false);
              }, Math.floor(flushInterval/2) + 100);
            } else {
              lokger(syslog.LOG_CRIT, "Error connecting to Librato!\n" + errdata);
            }
        });
      }

      var concurrent_conns = config.maxConnections || 10;
      var submissionq = async.queue(function (task,cb){
          task();
          cb();
          },concurrent_conns);

      _.each(combined_hashes,function(hash){
        _.each(_.keys(hash), function(hashtype) { /* 'gauges' and 'counters' */
          if (graphServiceIs("librato-metrics")){
            var stats_str = build_string(hash_postprocess(hash),hashtype)
            submissionq.push(function(){
              if (config.debug) {
                logger(syslog.LOG_DEBUG, stats_str);
                logger(syslog.LOG_DEBUG, stats_str.length);
              }
              submit_to_librato(stats_str,true);
            }, logerror);
          } else {
            var stats_str = build_string(hash_postprocess(hash),hashtype)
            submissionq.push(function(){
              if (config.debug) {
                logger(syslog.LOG_DEBUG, stats_str);
                logger(syslog.LOG_DEBUG, stats_str.length);
              }

              try {
                var graphite = net.createConnection(config.graphitePort, config.graphiteHost);
                graphite.addListener('error', function(connectionException){
                  if (config.debug) {
                    logger(syslog.LOG_CRIT, connectionException);
                  }
                });
                graphite.on('connect', function() {
                  this.write(stats_str);
                  this.end();
                  globalstats['graphite']['last_flush'] = Math.round(new Date().getTime() / 1000);
                  });
              } catch(e){
                if (config.debug) {
                  logger(syslog.LOG_DEBUG, e);
                }
              }
            }, logerror);
          }
        });
      });
    }, flushInterval);
  }
});
